### ISSUE TITLE
[sdl2] Change BDirectWindow to BWindow

### ISSUE THREAD

**Issue by X547** [haikuports/haikuports#4457]
BDirectWindow is not working properly. It has a lot of glitches such as flickering, cursor erasing, artifacts when there are windows on BDirectWindow. Using BWindow and BBitmap should not cause performance issues on post 2005 year PC. OpenGL software rasterizer is much more resouce consuming that drawing BBitmap. When hardware rendering become available BBitmap drawing in app_server can be done in GPU.

Current BDirectWindow implementation makes software such as Blender unusable.

**Comment by pulkomandy** [comment]
The artifacts should be fixed on Haiku side so that BDirectWindow just works.
Wether it is appropriate for SDL2 or not is an open question however, but it is unrelated. Either SDL2 benefits from direct access to the framebuffer and we should make this work, or it needs a backbuffer anyway for its drawing and in that case we may as well use an offscreen bitmap and SetViewBitmap() for rendering (this was benchmarked to be more efficient than DrawBitmap calls, and is what is used for example in the native port of MilkyTracker).

**Comment by X547** [comment]
Current sdl2 implementation uses BBitmap for buffer and use BDirectWindow to display it (SDL_BWin.h, SDL_bframebuffer.cc). BDirectWindow is meaningless in such usage.

**Comment by X547** [comment]
It is not possible to fix BDirectWindow without hardware support (hardware cursor, vsync), unless some kind of emulation used, for example using BBitmap as framebuffer and draw it to window by timer, emulating BScreen.WaitForRetrace to synchronize with timer. BeOS do not allow non-fullscreen BDirectWindow when no hardware support available.

**Comment by pulkomandy** [comment]
Our implementation is very different from the one in BeOS, as it does not allow access to the graphics card memory, only to app_server backbuffer. And it is not synchronized to the vblank (none of app_server drawing is). So, it could be made to work better. The cursor flickering may be a problem, but we can solve that too, for example by redrawing the cursor in app_server right before copying the backbuffer to the video card.

**Comment by X547** [comment]
> Our implementation is very different from the one in BeOS, as it does not allow access to the graphics card memory, only to app_server backbuffer. And it is not synchronized to the vblank (none of app_server drawing is).
In such case how app_server will decide when to copy back buffer part managed by BDirectWindow to front buffer?

**Comment by X547** [comment]
Timing is important for single buffer rendering. Consider loop:
```C++
while (run) {
    Draw(frontBuf);
    VSync();
}
```
If vsync is not working then some random immediate drawings will be displayed on screen causing flickering.

**Comment by pulkomandy** [comment]
I know, but this is not how app_server works currently. So whatever you try, you will get tearing. We should fix that, but this is better discussed on the Haiku bugtracker :)

**Comment by X547** [comment]
> Our implementation is very different from the one in BeOS, as it does not allow access to the graphics card memory, only to app_server backbuffer.

No. Haiku behavior is same as BeOS. I investigated source code and found that BDirectWindow uses front buffer. https://git.haiku-os.org/haiku/tree/src/servers/app/ServerWindow.cpp#n4191

> So whatever you try, you will get tearing.
There is difference between flickering and tearing. Flickering is displaying immediate contents before drawing is finished. Tearing is displaying part of previous frames when displaying current frame.

I made test application and managed to display contents (overlapping rectangles) on BDirectWindow without flickering by using BScreen.WaitForRetrace.

> We should fix that, but this is better discussed on the Haiku bugtracker :)
Is there existing ticket for that?
